Increased misconception allowed size from 30 to 60 characters

Showtim3 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #6481 
This is a pretty simple issue and all i did was increased the maxlength of the input from 30 to 60 as discussed on the issue thread. And accordingly I changed the error message too. Nothing's major has changed here. Screenshot Attached.

![misconception](https://user-images.githubusercontent.com/30312043/60682561-568d1f80-9eb1-11e9-8531-ecdf9acdb609.png)



## Checklist
- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [X] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [X] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [X] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

Showtim3 committed:
Increased misconception allowed size from 30 to 60 characters

Showtim3 commented:
@seanlip @aks681 hey PTAL, i didn't knew who else to tag.
Thanks!

seanlip commented:
@aks681 could you PTAL?

Also, just to check., are any changes needed to the backend validation? Or, @Showtim3, were you able to fill in and submit/save a 60-char misconception successfully? (We should generally be having parity between the restrictions in the frontend and the restrictions in the backend, so if it's not there already, we should probably add this as a validation check in the backend.)

aks681 commented:
I couldn't find a backend validation that restricted the length.
Though, @Showtim3 can you confirm that you can actually save a misconception of >30 length?
If so, I'd be happy to approve it!

seanlip commented:
Hi @aks681 -- just to check, we _should_ be adding a backend validation, right?

aks681 commented:
Hmm, this is just an HTML validation, right? I don't think any such validations have corresponding backend validations. (topic name, story title etc.)

seanlip commented:
I guess it depends on how strongly we want to enforce it. E.g. exploration titles are validated to be up to 50 chars (see utils.require_valid_name which is called by Exploration.validate()).

A validation check helps preserve invariants in the datastore models, so if we want to be absolutely certain in the future that all the stuff in the datastore is formatted a certain way, I'd suggest adding the backend check now rather than later (and to other places, as needed).

Showtim3 commented:
@aks681 
@seanlip 
Ok so the misconceptions are being saved also. 
Can you please give me some pointers on where should the corresponding backend validations should be made? 
![misconception saved](https://user-images.githubusercontent.com/30312043/60700138-5f630d00-9f14-11e9-97b1-105581909582.png)

seanlip commented:
@aks681 can give you more guidance, but check out the validate() methods in core.domain.skill_domain.py. Make sure to write tests!

aks681 commented:
Yeah, you can check out the misconception class in skill_domain and add a validation check to the existing validate function.

DubeySandeep commented:
@aks681, Do we have to make sure that the existing skills in the test server pass the new validation check? (Assuming that we don't have any skills in the production server.)

seanlip commented:
I think I'm going to try clearing the test and production servers of the new structures stuff, so we can start afresh there. (Probably easier.)

Showtim3 committed:
Added backend validation for misconception name length

Showtim3 committed:
Fixing linting

Showtim3 committed:
Fixed linting

Showtim3 commented:
@aks681 @seanlip hey guys this is my first backend contibution in Oppia and i am not really proficient at testing but I tried my best. 

So i have manually tested the backend validation. For that i removed the text input limit and added a misconception name greater than 60 chars and hence got an error from the backend. So i guess the backend validation is working as expected.

![backend](https://user-images.githubusercontent.com/30312043/60771399-48374180-a105-11e9-8a9c-4dad5d61bafa.png)

PTAL and i am open to any crticism or changes in my code.
Thanks!

DubeySandeep commented on core/domain/skill_domain.py:
```diff
@@ -246,6 +246,10 @@ def validate(self):
             raise utils.ValidationError(
                 'Expected misconception name to be a string, received %s' %
                 self.name)
+        if len(self.name) > 60:
```
Can we use `util.require_valid_name` here?

DubeySandeep commented on core/domain/skill_domain_test.py:
```diff
@@ -69,6 +69,14 @@ def test_valid_misconception_id(self):
             'Expected misconception ID to be an integer')
 
+    def test_valid_misconception_name(self):
+        name = ('Lets generate a string of more than 60 characters'
```
This looks great, we can make it better:

Do a baseline verification, i.e, len(name) < 50 passes the validation check after that change name such that len(name) > 50 and this time validation should fail. (see [here](https://github.com/oppia/oppia/wiki/Writing-backend-tests) for more info)

DubeySandeep commented on core/domain/skill_domain_test.py:
```diff
@@ -69,6 +69,14 @@ def test_valid_misconception_id(self):
+        self._assert_validation_error(
+            'Expected misconception name to be less than or equal to 60')
+
+
```
Only one newline required. 

(@oppia/dev-workflow-team I was expecting a lint test to catch this, is that correct?)

DubeySandeep reviewed (commented):
@Showtim3, Thanks for adding the tests, it looks great. :)

I've left a few comments PTAL!

Commit:
Added baseline verification, removed unnecessary newline, made code more generic

Commit:
Fixed linting

Commit:
Fixed linting

Commit:
Fixed linting

Commit:
Fixed linting

Showtim3 commented:
hey @DubeySandeep PTAL,
Thanks

DubeySandeep commented on core/domain/skill_domain.py:
```diff
@@ -246,6 +261,8 @@ def validate(self):
             raise utils.ValidationError(
                 'Expected misconception name to be a string, received %s' %
                 self.name)
+        self.require_valid_name(self.name)
```
Ah, I was asking you to check whether we can we use utils.require_valid_name over here? ([ref.](https://github.com/oppia/oppia/blob/develop/core/domain/exp_domain.py#L601))

Showtim3 replied:
No. Because in utils.require_valid_name throws an error if the string length is > 50 whereas our misconception name can be upto 60 chars.

DubeySandeep replied:
@Showtim3 I think we had finalized it to do 60 chars ([ref](https://github.com/oppia/oppia/issues/6481#issuecomment-508200257)).

DubeySandeep replied:
Ah, I miss understood your last comment! (Ignore my last comment)

DubeySandeep commented on core/domain/skill_domain_test.py:
```diff
@@ -69,6 +69,15 @@ def test_valid_misconception_id(self):
+    def test_valid_misconception_name(self):
+        name = 'A valid misconception name'
+        self.skill.update_misconception_name(0, name)
+        name = ('Lets generate a string of more than 60 characters'
```
Let's call self.skill.validate() here, just to do the verification that the name with length < 50 passes the validation check.

DubeySandeep commented on core/domain/skill_domain_test.py:
```diff
@@ -69,6 +69,15 @@ def test_valid_misconception_id(self):
             'Expected misconception ID to be an integer')
 
+    def test_valid_misconception_name(self):
+        name = 'A valid misconception name'
```
Let's rename the variable to misconception_name

DubeySandeep reviewed (commented):
@Showtim3, I've left a few a comments PTAL!

In general, after resolving a comment we use to add a reply to the comments stating it's "Done!", this makes it easier for the reviewer to do the review. (As stated [here](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change) in the 5th bullet) 
![image](https://user-images.githubusercontent.com/16653571/60828570-5ef69a80-a1d0-11e9-9c8c-3287074c3ec1.png)


Also, feel free to assign me (or any other reviewer) to this PR once it's ready for the review. ([ref.](https://groups.google.com/forum/#!topic/oppia-dev/Gg35jAKBIWw)), this will help us to quickly review your PR and get it merged! :)

Showtim3 committed:
Merged upstream develop

Showtim3 committed:
Fixed linting

Showtim3 committed:
Fixed linting #2

Showtim3 commented:
@DubeySandeep I'll just delete this pull request and open up a new one.
Can i do that?

seanlip commented:
Hi @Showtim3, yes you can, please go ahead!

Showtim3 commented:
Okay I'll be closing this PR and opening up a new one.

DubeySandeep commented:
> @DubeySandeep I'll just delete this pull request and open up a new one.
Can i do that?

(Just to update this PR thread) We had a discussion on (hangout) this and tried resolving the git issue and finalized that @Showtim3, will close this PR after creating a new PR (And add a reference to the new PR over here).

The new PR is at #7148.
